lib/pos_blockchain.go

```diff
@@ -0,0 +1,1331 @@
+		return false, nil
+	}
+	return true, nil
+}
+
+// isValidPoSQuorumCertificate validates that the QC of this block is valid, meaning a super majority
+// of the validator set has voted (or timed out). Assumes ValidatorEntry list is sorted.
+func (bc *Blockchain) isValidPoSQuorumCertificate(block *MsgDeSoBlock, validatorSet []*ValidatorEntry) error {
+	validators := toConsensusValidators(validatorSet)
+	if !block.Header.ValidatorsTimeoutAggregateQC.isEmpty() {
+		if !consensus.IsValidSuperMajorityAggregateQuorumCertificate(block.Header.ValidatorsTimeoutAggregateQC, validators) {
+			return RuleErrorInvalidTimeoutQC
+		}
+		return nil
+	}
+	if !consensus.IsValidSuperMajorityQuorumCertificate(block.Header.ValidatorsVoteQC, validators) {
+		return RuleErrorInvalidVoteQC
+	}
+	return nil
+}
+
+// getLineageFromCommittedTip returns the ancestors of the block provided up to, but not
+// including the committed tip. The first block in the returned slice is the first uncommitted
+// ancestor.
+func (bc *Blockchain) getLineageFromCommittedTip(block *MsgDeSoBlock) ([]*BlockNode, error) {
+	highestCommittedBlock, idx := bc.getCommittedTip()
+	if idx == -1 || highestCommittedBlock == nil {
+		return nil, errors.New("getLineageFromCommittedTip: No committed blocks found")
+	}
+	currentHash := block.Header.PrevBlockHash.NewBlockHash()
+	ancestors := []*BlockNode{}
+	prevHeight := block.Header.Height
+	prevView := block.Header.ProposedInView
+	for {
+		currentBlock, exists := bc.blockIndexByHash[*currentHash]
+		if !exists {
+			return nil, RuleErrorMissingAncestorBlock
+		}
+		if currentBlock.Hash.IsEqual(highestCommittedBlock.Hash) {
+			break
+		}
+		if currentBlock.IsCommitted() {
+			return nil, RuleErrorDoesNotExtendCommittedTip
+		}
+		if currentBlock.IsValidateFailed() {
+			return nil, RuleErrorAncestorBlockValidationFailed
+		}
+		if currentBlock.Header.ProposedInView >= prevView {
+			return nil, RuleErrorParentBlockHasViewGreaterOrEqualToChildBlock
+		}
+		if uint64(currentBlock.Header.Height)+1 != prevHeight {
+			return nil, RuleErrorParentBlockHeightNotSequentialWithChildBlockHeight
+		}
+		ancestors = append(ancestors, currentBlock)
+		currentHash = currentBlock.Header.PrevBlockHash
+		prevHeight = currentBlock.Header.Height
+		prevView = currentBlock.Header.ProposedInView
+	}
+	return collections.Reverse(ancestors), nil
+}
+
+// getOrCreateBlockNodeFromBlockIndex returns the block node from the block index if it exists.
+// Otherwise, it creates a new block node and adds it to the blockIndexByHash and blockIndexByHeight.
+func (bc *Blockchain) getOrCreateBlockNodeFromBlockIndex(block *MsgDeSoBlock) (*BlockNode, error) {
+	hash, err := block.Header.Hash()
+	if err != nil {
+		return nil, errors.Wrapf(err, "getOrCreateBlockNodeFromBlockIndex: Problem hashing block %v", block)
+	}
+	blockNode := bc.blockIndexByHash[*hash]
+	if blockNode != nil {
+		return blockNode, nil
+	}
+	prevBlockNode := bc.blockIndexByHash[*block.Header.PrevBlockHash]
```

Comment:
I want to point out here that prevBlockNode would be nil in the case where you fetch blocks in a suboptimal order. For example:
* Imagine the chain is A -> B -> C
* The node currently has A only, and now receives C
* We would create a BlockNode for C and add it to our index with Parent = nil

I'm mentioning the above because we need to later FIX C after we receive B. I'm still early in the code, but making a note to make sure we're doing this.

Reply:
This is broken even in PoW. There are a handful of cases where Parent is not set. The safe way to access a parent is to grab the parent's hash from the Header and look it up from the block index.